src/diator/middlewares/base.py

```diff
@@ -1,36 +1,37 @@
-from typing import Protocol
+import functools
+from typing import Protocol, TypeVar, Callable, Awaitable
 from diator.requests.request import Request
 
 from diator.requests.request_handler import RequestHandler
 from diator.response import Response
 
 
+Req = TypeVar("Req", bound=Request, contravariant=True)
+Res = TypeVar("Res", Response, None, covariant=True)
+HandleType = Callable[[Req], Awaitable[Res]]
+
+
 class Middleware(Protocol):
-    @property
-    def request_handler(self) -> RequestHandler:
+    async def __call__(self, request: Request, handle: HandleType) -> Res:
         ...
 
-    @request_handler.setter
-    def request_handler(self, request_handler: RequestHandler) -> None:
-        ...
 
-    async def process_request(self, request: Request) -> None:
-        ...
+class MiddlewareChain:
+    def __init__(self) -> None:
+        self._chain: list[Middleware] = []
 
-    async def process_response(self, response: Response) -> None:
-        ...
+    def set(self, chain: list[Middleware]) -> None:
+        self._chain = chain
 
+    def add(self, middleware: Middleware) -> None:
+        self._chain.append(middleware)
 
-class BaseMiddleware:
-    def __init__(self) -> None:
-        self._request_handler: RequestHandler | None = None
+    def wrap(self, handler: RequestHandler[Req, Res]) -> RequestHandler[Req, Res]:
+        handle = handler.handle
+
+        for middleware in reversed(self._chain):
+            handle = functools.partial(middleware.__call__, handle=handle)  # type: ignore
 
-    @property
-    def request_handler(self) -> RequestHandler:
-        if not self._request_handler:
-            raise RuntimeError("RequestHandler is not set.")
-        return self._request_handler
+        handler.handle = handle  # type: ignore
```

Comment:
Why is `handler.handle` overwritten here? 
If the state of the `RequestHandler` class is saved between calls, wouldn't `handler.handle` get rewritten each time with new middlewares?
Wouldn't it be better to make...?
```python
handle = partial(handler.handle, handler=handle)
return handle
```

Reply:
@SamWarden yes, seemingly this logic can be simplified

Reply:
@SamWarden now `wrap` method gets `handle` directly